res/controllers/Denon-Prime-4-scripts.js

```diff
@@ -0,0 +1,879 @@
+        },
+        inputWheel: function(channel, control, value, _status, _group) {
+            value = this.relativeFromAbsolute(value);
+            if (engine.isScratching(this.deck)) {
+                engine.scratchTick(this.deck, value);
+            } else {
+                this.inSetValue(this.jogScale(value));
+            }
+        },
+    });
+
+    // Jog Wheel LED
+    this.jogWheelLed = new components.Component({
+        midi: [0x90 + midiChannel, 0x21],
+        on: colDeck[deckNumbers - 1],
+        trigger: function() {
+            this.send(this.on);
+        },
+    });
+
+    // Slip Mode Button
+    this.slipButton = new components.Button({
+        midi: [0x90 + midiChannel, 0x24],
+        key: "slip_enabled",
+        type: components.Button.prototype.types.toggle,
+    });
+
+    // Loop In Button
+    this.loopInButton = new components.Button({
+        midi: [0x90 + midiChannel, 0x25],
+        key: "loop_in",
+    });
+
+    // Loop Out Button
+    this.loopOutButton = new components.Button({
+        midi: [0x90 + midiChannel, 0x26],
+        key: "loop_out",
+    });
+
+    // Load Buttons
+    this.deckLoad = new components.Button({
+        midi: [0x9F, midiChannel - 3],
+        key: "LoadSelectedTrack",
+
+        // TODO: Make load buttons reflect colour of currently active deck.
+        //       This is harder than expected due to the fact that I rely on the MIDI channel
+        //       of each physical deck component to determine which deck it's manipulating in
+        //       Mixxx. The deck load buttons, however, are both on MIDI channel 16, so I need
+        //       a better way to determine this.
+        off: Prime4.rgbCode.whiteDark,
+        on: Prime4.rgbCode.white,
+    });
+
+    this.reconnectComponents(function(c) {
+        if (c.group === undefined) {
+            c.group = this.currentDeck;
+        }
+    });
+};
+
+Prime4.Deck.prototype = new components.Deck();
+
+Prime4.shift = false;
+Prime4.shiftState = function(channel, control, value) {
+    Prime4.shift = value === 0x7F;
+    if (Prime4.shift) {
+        midi.sendShortMsg(0x90 + channel, control, 0x02);
+        Prime4.leftDeck.shift();
+        Prime4.rightDeck.shift();
+        Prime4.effectBank[0].shift();
+        Prime4.effectBank[1].shift();
+    } else {
+        midi.sendShortMsg(0x90 + channel, control, 0x01);
+        Prime4.leftDeck.unshift();
+        Prime4.rightDeck.unshift();
+        Prime4.effectBank[0].unshift();
+        Prime4.effectBank[1].unshift();
+    }
+};
+
+//========== PERFORMANCE PADS ==========//
+
+// Access the appropriate mode-select pad without remembering MIDI values
+Prime4.padMode = {
+    HOTCUE: 0x0B,
+    LOOP: 0x0C,
+    ROLL: 0x0D,
+    SLICER: 0x0E,
+};
+
+Prime4.PadSection = function(deck, offset) {
+    components.ComponentContainer.call(this);
+
+    // Create component containers for each pad mode
+    this.modes = new components.ComponentContainer({
```

Comment:
While looking at the debug log from you, I noticed an issue where pads where being disconnected twice, this should fix that. 
```suggestion
    const modes = new components.ComponentContainer({.
```

``javascript
const controlToPadMode = control => {
    // copy paste controlToPadMode from below here instead use `modes` instead of `this.modes`
};
```

Reply:
Getting this error now after doing that:
```
Uncaught exception: file:///home/whanake/.mixxx/controllers/Denon-Prime-4-scripts.js:724: TypeError: Property 'controlToPadMode' of object [object Object] is not a function
Backtrace: @file:///home/whanake/.mixxx/controllers/Denon-Prime-4-scripts.js:724
@file:///home/whanake/.mixxx/controllers/Denon-Prime-4-scripts.js:756
@file:///home/whanake/.mixxx/controllers/Denon-Prime-4-scripts.js:512
@file:///home/whanake/.mixxx/controllers/Denon-Prime-4-scripts.js:267
```

Reply:
Yeah, you'll also need to change the locations where its invoked. So in this case the issue stems from the fact that you forgot to change to `const newMode = controlToPadMode(control);` in `setPadMode`.

Reply:
Ah I see. Thanks, working again. Still clashing pad modes though. Pushing changes now.

Reply:
Thank you. I guess for some reason, the padmodes get connected twice but I'm not sure why or when. Can you provide me with a full log from mixxx startup? You can look into the `.mixxx` folder for it if its too long for your terminal.

Reply:
Log is available at `mixxx/res/controllers/StartupLog.txt`